### ISSUE TITLE
API to wrap existing layers and make them use parameters lying on a manifold

### ISSUE THREAD

**Issue by ferrine** [ferrine/geoopt#5]
Need to decide on a good intuitive API to make for example orthogonal Conv2d or so. Any suggestions are welcome.

```python
conv = nn.Conv2d(...)
conv = geoopt.make_manifold(conv, name='weight', manifold=geoopt.Stiefel())
```

**Comment by justanothercoder** [comment]
We can make it as in [weight_norm](https://pytorch.org/docs/stable/_modules/torch/nn/utils/weight_norm.html): delete old `nn.Parameter`, re-register new `ManifoldParameter`. Also, we need forward pre-hook because, for example, `Stiefel` is defined only for 2d-tensors. So, probably we need to transform the weight somehow before using it. 
For orthogonal conv it makes sense to have parameter of shape `[out_channels, in_channels * width * height]` and then reshape it to the `[out_channels, in_channels, width, height]` before conv.

For convs and FCs it may be easier just to create new layers.

**Comment by ferrine** [comment]
In the original paper another regularization scheme is shown to be useful
https://github.com/ajbrock/Neural-Photo-Editor/blob/master/train_IAN.py#L158
there authors regularize filters to be orthogonal, so they impose Stiefel manifold on 2 last dimensions of convolutional kernel tensor. 
Therefore it makes sense to support broadcasted manifolds for the last dims as well. 

Do I correctly understand that this setting is indeed allows to treat manifolds independently?

**Comment by ferrine** [comment]
The approach as in weight norm seems to be a good practice, we can take the best of it.

**Comment by justanothercoder** [comment]
> Do I correctly understand that this setting is indeed allows to treat manifolds independently?

Yes.

**Comment by ferrine** [comment]
```python
class OrthConv2d(nn.Conv2d):
    def __init__(self, *args, alpha=None, **kwargs):
        self._weight = None
        self._weight_shape = None
        super().__init__(*args, **kwargs)
        weight = self._parameters.pop("weight")
        self._weight_shape = weight.shape
        self.weight_orig = geoopt.ManifoldParameter(
            weight.data.reshape(weight.shape[0], -1), manifold=geoopt.Stiefel()
        )
        self.weight_orig.proj_()
        if alpha is None:
            self.alpha_orig = nn.Parameter(torch.zeros(self._weight_shape[0]))
        else:
            self.alpha_orig = alpha

    @property
    def weight(self):
        return (self.alpha[:, None] * self.weight_orig).reshape(self._weight_shape)

    @property
    def alpha(self):
        return self.alpha_orig.exp()
```

**Comment by ferrine** [comment]
There is yen no dim specification, we can think about it
